### ISSUE TITLE
[misc-definitions-in-headers] incorrectly warns definition in unnamed namespace

### ISSUE THREAD

**Issue by namniav** [llvm/llvm-project#61471]
This was first reported in clangd/clangd#1543 and I was told that it should be reported here.

Clang incorrectly warns definition in unnamed namespace in header file.
```c++
// test.hpp
namespace a {
  namespace  /*b*/  {
    static int c[[maybe_unused]] = 42; // even static doesn't help
  }
}
```
```console
test.hpp|3 col 16 warning| Variable 'c' defined in a header file; variable definitions in header files can lead to ODR violations [misc-definitions-in-headers]
```
However, this only happens in unnamed namespace. If I uncomment `b`, the warning is gone.

**System information**

Output of `clang --version`:
```console
Homebrew clang version 15.0.7
Target: arm64-apple-darwin21.4.0
Thread model: posix
```
Editor/LSP plugin: Vim 9.0 + YouCompleteMe
Operating system: macOS

**Comment by namniav** [comment]
The code demo above should be sufficient. All that I do is just open `test.hpp` in vim with `-Wall -Wextra -Wpedantic` configurated to ycm/clangd.
But I don't know how to reproduce this warning with clang command line, so I can't provide a godbolt link. I would appreciate it if someone could give me the command line options.

**Comment by namniav** [comment]
Oh, I just figured it out: `clang-tidy -checks=misc-definitions-in-headers`.

```c++
❯ cat test.hpp
namespace a {
  namespace   {
    static int c[[maybe_unused]] = 42; // even static doesn't help either
  }
}
❯ clang-tidy -checks=misc-definitions-in-headers test.hpp
1 warning generated.
/Users/vainman/test/test.hpp:3:16: warning: variable 'c' defined in a header file; variable definitions in header files can lead to ODR violations [misc-definitions-in-headers]
    static int c[[maybe_unused]] = 42; // even static doesn't help either
```

**Comment by namniav** [comment]
https://github.com/llvm/llvm-project/blob/62f09d788f9fc540db12f3cfa2f98760071fca96/clang-tools-extra/clang-tidy/misc/DefinitionsInHeadersCheck.cpp#L97-L100

I think this can be fixed by simply removing `&& !ND->isInAnonymousNamespace()` ?

**Comment by namniav** [comment]
Yes, warning variables with internal linkage in headers is also a reasonable choice. Although your linked example seems not odr-use `pi`, I can be wrong since I'm not familiar with standard terms, and please correct me if I'm wrong.

I was thinking that current behavior seems be intended to warn only non-inline variables with external linkage in headers, so not to warn variables with internal linkage is consistent with current behavior, since they are fine as long as they are not odr-used in entity with external linkage. The comment in code seems somehow confirm my guess. Anyway, current behavior is not consistent.

**Comment by carlosgalvezp** [comment]
> I can be wrong since I'm not familiar with standard terms, and please correct me if I'm wrong.
Being pedantic about the standard, the following would be UB:
```cpp
namespace {
int x = 123;
}

inline int foo()
{
  return x * 2;
}
```
Because in reality, each translation unit has:
```cpp
inline int foo()
{
  return some_unique_namespace_for_this_tu::x  * 2;
}
```

Therefore the function `foo` is not "token by token" identical in each TU.

Anyway, reading the docs of this check I understand the goal is to handle "the typical case" and ignore the more contrived ones (like e.g. internal linkage via `static` or `const`, which is similar to using anonymous namespaces). We can have more pedantic checks for the more contrived examples so people can enable if wanted.

**Comment by carlosgalvezp** [comment]
Patch: https://reviews.llvm.org/D147379

**Comment by namniav** [comment]
> Although your linked example seems not odr-use `pi`, I can be wrong since I'm not familiar with standard terms, and please correct me if I'm wrong.

I meant that `pi` is `const` in linked SO example, so `twoPiR` doesn't [odr-use](https://stackoverflow.com/questions/19630570) `pi`, does it?

Thank you for the patch.

**Comment by carlosgalvezp** [comment]
Yes exactly, it does not ODR-use it, sorry if that was unclear. It's nevertheless UB according to the Standard (whether it's actually a problem or not in practice is another topic :) )
